gobblin-api/src/main/java/gobblin/source/extractor/Extractor.java

```diff
@@ -74,4 +83,38 @@
    */
   @Deprecated
   public long getHighWatermark();
+
+  /**
+   * Read an {@link RecordEnvelope}. By default, just wrap {@link #readRecord(Object)} in a {@link RecordEnvelope}.
+   */
+  default RecordEnvelope<D> readRecordEnvelope() throws DataRecordException, IOException {
+    D record = readRecord(null);
+    return record == null ? null : new RecordEnvelope<>(record);
+  }
+
+  /**
+   * @param shutdownRequest an {@link AtomicBoolean} that becomes true when a shutdown has been requested.
+   * @return a {@link Flowable} with the records from this source. Note the flowable should honor downstream backpressure.
+   */
+  default RecordStreamWithMetadata<D, S> recordStream(AtomicBoolean shutdownRequest) throws IOException {
```

Comment:
I though about that, but having a `shutdown` method is hard because interfaces cannot hold instance variables, so the shutdown has no way of notifying the stream to stop.